### ISSUE TITLE
(Generic)OAuth2 API refresh_token is not used with teamleader.eu -- 2nd Try

### ISSUE THREAD

**Issue by nesretep-anp1** [n8n-io/n8n#5623]
**Describe the bug**
We try to use TeamLeader.eu API via N8N and created an OAUTH2 API credential in N8N which works fine (account is connected). Appropriate requests to the API are working well for around 60 minutes. Then the error "access token expired" is issued. 

**To Reproduce**

1. Register an OAuth client on TeamLeader.eu
2. Create the credential in N8N (we tried both, Header and Body)
3. Connect works well.
4. Having a workflow with HTTP Request and Authentication => Generic, Generic Auth Type => OAUTH2 API.
5. Workflow works well for around 60 minutes, then "access token expired".

**Expected behavior**
It is proven that TeamLeader.eu API returns a refresh_token from the very beginning. So the expected behaviour is that there should not be an expiry due to having the refresh_token being used to get a new access_token.

**Environment (please complete the following information):**

- Ubuntu Linux 22.04
- n8n Version 0.217.2
- Node.js Version 14.21.3
- Database system Postgres
- Operation mode main

**Additional context**
#5607
Unfortunately the given issue was closed without collecting additional infos. Like said in the given comment there (and already mentioned in Reproduction 4.): we do use HTTP Request Node. But all configuration parameters are set as documentation, hints/tips and appropriate sources state.

IF this is NOT a bug IN the software, THEN it IS a BUG in documentation!

Yes, we already tried N8N_USE_DEPRECATED_REQUEST_LIB env variable.

**Comment by Joffcom** [comment]
Hey @nesretep-anp1 

The previous issue was closed because we typically only use GitHub issues for bugs.

Based on what you have provided this could be that something is missing in the api scope, I am not aware of the api you are using and I have not seen the documentation for it but typically this sort of request would start on the support forum.

Can you provide a link to the api doc you are following and share what settings you are using in the oauth credential and I can take a look.

**Comment by nesretep-anp1** [comment]
Hey @Joffcom 

thanks for adopting.

The API docs could be found here: https://developer.teamleader.eu/#/introduction/authentication

Like said, we already tried with and without N8N_USE_DEPRECATED_REQUEST_LIB (which is discussed in the community forums); like we already checked out the other additional params given there. (Like e.g. switch from HEADER to BODY:)

As seen in the API doc, the scope parameter on auth is not needed(/used) as this is given on registration of the client app.

So, ... on the one hand we already used all given settings out of community forums and on the other hand, like stated in docs and by your stuff, refresh token should work out of the box; so, ... this is why I say, that this is a bug - software or docs. ;)

![teamleader-creds](https://user-images.githubusercontent.com/47453303/223060375-d7cc46ac-597d-4241-b7d5-3d6e0c3fcf86.png)

(The screenshot is the current state; like said, we already tried many differen combinations).

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

I have set up a quick trial account and configured the options that I think need to be set, I will be back in an hour to find out if it still works.

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

Currently it looks like my workflow is running as expected.

![image](https://user-images.githubusercontent.com/4688521/223105536-7041836d-4eb5-4aee-91a0-bc46dfa307b8.png)

For my test I have been calling `https://api.focus.teamleader.eu/users.me` although I suspect that part doesn't really matter, I have just updated my schedule to every 2 hours just to make sure there is over 60 minutes between runs but initial testing would suggest all is good.

**Comment by nesretep-anp1** [comment]
> For my test I have been calling https://api.focus.teamleader.eu/users.me

Well, yes, there some examples talking about that.

> although I suspect that part doesn't really matter,

That's what I think too; but - currently - a difference between your "on the fly" setup and ours!

> I have just updated my schedule to every 2 hours just to make sure there is over 60 minutes between runs but initial testing would suggest all is good.

Node => lts/fermium ?

**Comment by nesretep-anp1** [comment]
Nice one; that was exactly the first try I used upon starting the TeamLeader integration! (Except for using "projects" as scope.)
:thinking:

**Comment by nesretep-anp1** [comment]
What else could I say!?

```
{"body":{"errors":[{"title":"Access token expired","status":401}]},"headers":{"date":"Mon, 06 Mar 2023 14:30:00 GMT","content-type":"application/json","transfer-encoding":"chunked","connection":"close","x-trace-id":"01633c0962fbf3f379fea362a0ce89dd","access-control-allow-origin":"*","strict-transport-security":"max-age=15724800; includeSubDomains"},"statusCode":401,"statusMessage":"Unauthorized"}
```

**Comment by Joffcom** [comment]
For me with the 2 hour delay it has still worked,  Can you share what you have set for the credential at the moment?

**Comment by nesretep-anp1** [comment]
The following parameters are currently set (Remember, ... we already tested many several combinations) and resulted back in the given error:

Grant Type
=> Authorization Code
Authorization URL
=> https://focus.teamleader.eu/oauth2/authorize
Access Token URL
=> https://focus.teamleader.eu/oauth2/authorize
Client ID
=> _removed_
Client Secret
=> _removed_
Scope
=> _empty_
Auth URI Query Parameters
=> access_type=offline
Authentication
=> Header

**Comment by nesretep-anp1** [comment]
I do see a string "n8n_BLANK_VALUE" included in the /rest/oauth2-credential/auth calls when connecting the account!?

**Comment by Joffcom** [comment]
Hey @nesretep-anp1 

I would expect n8n blank value to be there if you have toggled the expression option for the secret.

I know said "remember we have tested several options" but given that I have managed to get this working on my first attempt I think it is safe to assume it is likely an issue with the options you have set.

You still don't have a scope set, but can I get a screenshot of the settings you have like I posted above so I can see them rather than a text write up as there are things that can be missed.

**Comment by nesretep-anp1** [comment]
If you need a new screenshot, .... no problem. But the fields look still very similiar to my first screenshot here in this thread!?

**Comment by Joffcom** [comment]
Hey @nesretep-anp1 

Yes please, as I can't see all the fields in your earlier screenshot and the values you sent in your text version don't match that screenshot so I want to make sure I get it exact.

For example in your text you mentioned having the Auth uri parameter set but in the screenshot you only have the placeholder text and no value.

I am also not sure why you are seeing n8n_blank_value so I suspect at some point maybe the secret was changed to an expression then back to a normal field which also could have broken it but that would be unusual.

Another option is if you delete the credential you have and put in what I have set including the Auth URI then try that and see if it works.

If you are still using deprecated request lib I would remove that as  well.

**Comment by nesretep-anp1** [comment]
Ok, will send the screenshot.

But, ... my thoughts are, that the empty "scope" does result in the blank value and destroy the refresh. I changed our creds to using a scope "users" this morning and now I'll wait if it still expires. Of course that's not usable for us, because we do need all scopes offered by teamleader, but we would be able to identify the bug.

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

I have created a new credential on mine that has no scope to see if anything changes, It would be mildly annoying if that is the cause although I know we have some nodes that use OAuth with no scopes set.

I will be back in 2 hours.

**Comment by nesretep-anp1** [comment]
:cry: Ok. No success even with scope. Access Token expired.

And, by the way, ... the account connect AGAIN let to using `__n8n_BLANK_VALUE_` in callback URL!?

IF you do not find any differences in the attached screenshots, then the only 2 things left are:

* lts/fermium
* main

![tl07-1](https://user-images.githubusercontent.com/47453303/223399996-5197672e-2c96-4d7c-8a5f-9212d948ebe8.png)
![tl07-2](https://user-images.githubusercontent.com/47453303/223400001-87a9c3b8-b3f4-4ea8-ab73-4e26afd71083.png)

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

I have just manually ran my workflow as we are over the 60 minute mark anyway and it is working as expected still.

Can you maybe share the callback URL you are seeing with the blank value in it? I can't see what you are seeing so the more you can share the better. 

Main shouldn't be an issue as it is what I am using along with Postgres so the only other difference at the moment is I am using the recommended node version.

Can you also share the workflow you are running? Maybe there is something else I am missing that could be relevant.

**Comment by nesretep-anp1** [comment]
Ok, ... first find attached the according test workflow.

Second, ... the URL called while the last account connect:

`GET /rest/oauth2-credential/auth?id=14&authUrl=https:%2F%2Ffocus.teamleader.eu%2Foauth2%2Fauthorize&accessTokenUrl=https:%2F%2Ffocus.teamleader.eu%2Foauth2%2Faccess_token&clientId=(_removed_)&clientSecret=__n8n_BLANK_VALUE_(_removed_BUT_LOOKED_LIKE_A_GUID_)&scope=users&authQueryParameters=access_type%3Doffline&oauthTokenData=__n8n_BLANK_VALUE_(_removed_BUT_LOOKED_LIKE_A_GUID_)&grantType=authorizationCode&authentication=header`

[retrieve_data_teamleader.json.txt](https://github.com/n8n-io/n8n/files/10908823/retrieve_data_teamleader.json.txt)

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

So that __n8n_BLANK_VALUE part is for the secret, Did you ever change the secret value to an expression then back to the other option?

**Comment by nesretep-anp1** [comment]
Not knowingly.

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

I have now set up your workflow to run to see if that changes anything, The only thing I can think of is at some point the credential was toggled from Expression to Fixed and that has changed the value but that is a change that you would know if you had done it as the field is normally masked. That and you made a new credential and unless you copied the old value from the credential that wouldn't be an issue.

Can you run the below and check that the value is correctly set in the output?

```
n8n export:credentials --id=14 --decrypted
```

**Comment by nesretep-anp1** [comment]
In the meantime I changed node to lts/gallium (which should correspond to your version). Funny thing is, that now the access token expires after 1 hour, not 2!?!? :cry: 

Beside that, ...

```
[{"createdAt":"2023-02-28T11:29:16.973Z","updatedAt":"2023-03-07T11:55:02.016Z","id":"14","name":"TeamLeader","data":{"authUrl":"https://focus.teamleader.eu/oauth2/authorize","accessTokenUrl":"https://focus.teamleader.eu/oauth2/access_token","clientId":"(_removed_)","clientSecret":"(_removed_)","scope":"users","authQueryParameters":"access_type=offline","oauthTokenData":{"token_type":"Bearer","expires_in":3600,"access_token":(_removed_was_VALID_JWT_)","refresh_token":"(_removed_)"}},"type":"oAuth2Api","nodesAccess":[{"nodeType":"n8n-nodes-base.graphql","date":"2023-02-28T11:29:50.865Z"}]}]
```

**Comment by nesretep-anp1** [comment]
And, by the way, ... if the client secret would get used with wrong content, would'nt it "destroy" the whole auth process and not only refresh?!

**Comment by nesretep-anp1** [comment]
Beside that, ... should `N8N_USE_DEPRECATED_REQUEST_LIB` be used here or not? (We had tried both, currently it is NOT used).

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

The credential looks ok assuming it didn't contain the n8n_blank_value, If the secret was wrong it should indeed stop all of it but at the moment it is the only thing I can think of. The use deprecated request lib shouldn't be set at all and it is something we don't really ask people to do anymore.

The test workflow I got from you is still running without issues, I am not sure what is going on as it is working fine for me, For a bit of fun I have set it up on a different n8n instance as well.

Can you tell me how you have n8n set up, How is it running, what environment options do you have set, is it the standard n8n version or a special custom version?

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

That is correct I also use main but I have 1 instance on own just to make sure that is ok as well, The credential is also still working as expected for me on 4 instances now. Do you have any other environment or config options set? I can't think of any that would impact this but from what I can see this appears to be environmental as I am not able to reproduce it.

You do only have the one instance right and you are not running in queue mode?

**Comment by Joffcom** [comment]
Can't see anything there that would cause an issue 🤔 and this is just the one instance of n8n running with nothing else. What were you using to view the auth requests?

**Comment by nesretep-anp1** [comment]
Is there something to enable/raise detail logs around the refresh stuff in n8n?!

**Comment by nesretep-anp1** [comment]
Found the bug!

![tl09-1](https://user-images.githubusercontent.com/47453303/224103644-d924c1d4-a5ee-4849-b37b-c20dc2f7f93d.png)

If these options are set, then refresh is never done! I did not test which one, ... but IMHO it is "Never error" which, if set, will never trigger via 401!

**Comment by Joffcom** [comment]
Now that is a fantastic find, that is something I had in the flow but didn't know why it was there so like a fool I removed it.

Will try and reproduce that properly in the morning and get the internal bug ticket created.

**Comment by Joffcom** [comment]
Alright little update... Today I have been running Include Response and Never Error and it wasn't failing. Just did some digging and I found that I had another workflow for the same credential running on a schedule without those options which was likely keeping the credential alive.

I have disabled the other workflow and only have the one going.

**Comment by nesretep-anp1** [comment]
Update from our side. Like in the workflow I sent in here, we do have an users.me before the things we really want to do with the API. Only the users.me (so the first HTTP request) has switched off both flags. This workflow is running for 3 days now.

But, ... IMHO this is a bug around refresh_token, Request HTTP node and OAuth2!?

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

Mine has started to fail as well and it does indeed look like a bug, When never fail is set we treat any response as a success which would be intended if it wasn't for the oauth process needing to happen so we will need a check for that.

The internal dev ticket for this one is `NODE-380`, Once fixed I will let you know.

**Comment by janober** [comment]
Fix got released with [`n8n@0.223.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%400.223.0)

**Comment by Joffcom** [comment]
Hey @nesretep-anp1,

This should now be resolved, I am going to mark this one as closed. If you still this issue let me know and we can open the issue again.
